crowbar_framework/lib/crowbar/repository.rb

```diff
@@ -0,0 +1,138 @@
+#
+# Copyright 2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  class Repository
+    attr_reader :platform, :id, :registry, :available, :md5, :url
+
+    #
+    # class methods
+    #
+    def self.load_registry
+      @@config ||= YAML.load_file(File.join(Rails.root, "config", "repos.yml"))
+    end
+
+    def self.where(options = {})
+      platform = options.fetch :platform, nil
+      repo = options.fetch :repo, nil
+      check_all_repos.select do |r|
+        if platform
+          if repo
+            r.platform == platform && r.name == repo
+          else
+            r.platform == platform
+          end
+        else
+          r.name == repo
+        end
+      end
+    end
+
+    def self.all_platforms
+      @@config.keys
+    end
+
+    def self.check_all_repos
+      repochecks = []
+      all_platforms.each do |platform|
+        repositories(platform).each do |repo|
+          check = self.new(platform, repo)
+          repochecks << check
+        end
+      end
+      repochecks
+    end
+
+    def self.admin_platform
+      NodeObject.admin_node.target_platform
+    end
+
+    def self.web_port
+      Proposal.where(barclamp: "provisioner").first.raw_attributes["web_port"]
+    end
+
+    def self.repositories(platform)
+      @@config[platform]["repos"].keys
+    end
+
+    #
+    # constructor
+    #
+    def initialize(platform, repo)
+      @platform = platform
+      @id = repo
+      @registry = Crowbar::Repository.load_registry[@platform]["repos"][@id]
+      @available = valid_repo_dir?
+      @md5 = valid_key_file?
+      @url = config_web_url
+    end
+
+    #
+    # instance methods
+    #
+    def valid_repo_dir?
+      all_repos_dir.include?(@registry["name"])
+    end
+
+    def valid_key_file?
+      key = key_file
+      if key
+        md5 = Digest::MD5.hexdigest(File.read(key))
+        config_repomd_md5.include? md5
+      else
+        false
+      end
+    end
+
+    private
+
+    #
+    # config
+    #
+    def config
+      @@config[@platform]["repos"][@id]
+    end
+
+    def config_repomd_md5
+      config["repomd"]["md5"]
+    end
+
+    def config_repomd_key
+      config["repomd"]["key"] || "/srv/tftpboot/#{@platform}/repos/#{@registry["name"]}/repodata/repomd.xml.key"
+    end
+
+    def config_web_url
```

Comment:
The `config_` prefix is not right here because if the URL is not defined in the config, it gets constructed by this method.  Also the URL might not be an HTTP one, so `web_` is also misleading.  So I think a better method name is simply `#url`.